### PR TITLE
Document custom configuration and env variables for Keycloak

### PR DESCRIPTION
**Pull request by TheBigLee**
This PR documents how to create and use a custom configuration and custom environment variables

### CODE REVIEW

**Review by Kidswiss — APPROVED** [review]

**Review by zugao — APPROVED** [review]
LGTM

File: docs/modules/ROOT/pages/vshn-managed/keycloak/customization.adoc [hunk 1]
```diff
@@ -32,3 +32,133 @@ spec:
 <1> Container image url.
 <2> Secret name containing credentials for the registry.
 <3> Secret namespace
+
+
+== Configuration
+
+You can provide a custom configuration to keycloak using a `ConfigMap``.
```
Comment (tobru) [inline docs/modules/ROOT/pages/vshn-managed/keycloak/customization.adoc:39]:
```suggestion
You can provide a custom configuration to Keycloak using a `ConfigMap`.
```

File: docs/modules/ROOT/pages/vshn-managed/keycloak/customization.adoc [hunk 2]
```diff
@@ -32,3 +32,133 @@ spec:
...
+The content of the `ConfigMap` is a JSON file based on the keycloak export files.
```
Comment (tobru) [inline docs/modules/ROOT/pages/vshn-managed/keycloak/customization.adoc:40]:
```suggestion
The content of the `ConfigMap` is a JSON file based on the Keycloak export files.
```

Do we have a link to somewhere which describes what a "Keycloak export file" is?

File: docs/modules/ROOT/pages/vshn-managed/keycloak/customization.adoc [hunk 3]
```diff
@@ -32,3 +32,133 @@ spec:
...
+The configuration *must* be in the key `keycloak-config.json` inside the `ConfigMap`
+
+The configuration will be read during the startup of keycloak and loaded into the database.
```
Comment (tobru) [inline docs/modules/ROOT/pages/vshn-managed/keycloak/customization.adoc:43]:
```suggestion
The configuration will be read during the startup of Keycloak and loaded into the database.
```

File: docs/modules/ROOT/pages/vshn-managed/keycloak/customization.adoc [hunk 4]
```diff
@@ -32,3 +32,133 @@ spec:
...
+
+An extensive set of configuration examples can be found https://github.com/inventage/keycloak-config-cli/tree/main/src/test/resources/import-files[here^].
+
+The configuration also supports variable substition.
```
Comment (tobru) [inline docs/modules/ROOT/pages/vshn-managed/keycloak/customization.adoc:47]:
```suggestion
The configuration also supports variable substitution.
```

File: docs/modules/ROOT/pages/vshn-managed/keycloak/customization.adoc [hunk 5]
```diff
@@ -32,3 +32,133 @@ spec:
...
+
+Variables exposed by spring boot (through configtree or external configuration) can be accessed by `$(property.name)`
```
Comment (tobru) [inline docs/modules/ROOT/pages/vshn-managed/keycloak/customization.adoc:49]:
```suggestion
Variables exposed by Spring Boot (through configtree or external configuration) can be accessed by `$(property.name)`
```

What is "configtree"? Are there any external links which describe these variables in more detail?

File: docs/modules/ROOT/pages/vshn-managed/keycloak/customization.adoc [hunk 6]
```diff
@@ -32,3 +32,133 @@ spec:
...
+
+In additional, the string substitution support multiple prefixes for different approaches
```
Comment (tobru) [inline docs/modules/ROOT/pages/vshn-managed/keycloak/customization.adoc:51]:
```suggestion
In additional, the string substitution support multiple prefixes for different approaches:
```

File: docs/modules/ROOT/pages/vshn-managed/keycloak/customization.adoc [hunk 7]
```diff
@@ -32,3 +32,133 @@ spec:
...
+URL Content (HTTPS):   $(url:UTF-8:https://www.apache.org)
+URL Content (File):    $(url:UTF-8:file:///$(sys:user.dir)/src/test/resources/document.properties)
+XML XPath:             $(xml:src/test/resources/document.xml:/root/path/to/node)
+----
+
+=== Example
+
+The following example demonstrates how a custom configuration that creates a new realm looks like:
+
+.Configmap definition for a custom realm to be created when starting up keycloak.
```
Comment (tobru) [inline docs/modules/ROOT/pages/vshn-managed/keycloak/customization.adoc:80]:
```suggestion
.ConfigMap definition for a custom realm to be created when starting up Keycloak
```

File: docs/modules/ROOT/pages/vshn-managed/keycloak/customization.adoc [hunk 8]
```diff
@@ -32,3 +32,133 @@ spec:
...
+[source,yaml]
+----
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: keycloak-app1-prod-config
+  namespace: bigli-test
```
Comment (tobru) [inline docs/modules/ROOT/pages/vshn-managed/keycloak/customization.adoc:87]:
```suggestion
  namespace: prod-app
```

File: docs/modules/ROOT/pages/vshn-managed/keycloak/customization.adoc [hunk 9]
```diff
@@ -32,3 +32,133 @@ spec:
...
+----
+apiVersion: vshn.appcat.vshn.io/v1
+kind: VSHNKeycloak
+metadata:
+  name: keycloak-app1-prod
+  namespace: prod-app
+spec:
+  parameters:
+    service:
+      customConfigurationRef: keycloak-app1-prod-config # <1>
```
Comment (tobru) [inline docs/modules/ROOT/pages/vshn-managed/keycloak/customization.adoc:107]:
```suggestion
      customConfigurationRef: keycloak-app1-prod-config <1>
```

File: docs/modules/ROOT/pages/vshn-managed/keycloak/customization.adoc [hunk 10]
```diff
@@ -32,3 +32,133 @@ spec:
...
+
+== Environment variables
+
+You can pass custom environment variables to your Keycloak instance. Those can then either be used by your custom providers or by your custom configuration.
+
+=== Example
+
+The following example demonstrates how a custom configuration that creates a new realm and uses environment variables looks like:
+
+.Secret containing a custom environment variable.
```
Comment (tobru) [inline docs/modules/ROOT/pages/vshn-managed/keycloak/customization.adoc:120]:
```suggestion
.Secret containing a custom environment variable
```

File: docs/modules/ROOT/pages/vshn-managed/keycloak/customization.adoc [hunk 11]
```diff
@@ -32,3 +32,133 @@ spec:
...
+kind: Secret
+metadata:
+  name: keycloak-app2-prod-env
+  namespace: prod-app
+stringData:
+  REALM_NAME: prod-app
+type: Opaque
+----
+
+.Configmap definition referencing an environmental variable.
```
Comment (tobru) [inline docs/modules/ROOT/pages/vshn-managed/keycloak/customization.adoc:133]:
```suggestion
.Configmap definition referencing an environmental variable
```

File: docs/modules/ROOT/pages/vshn-managed/keycloak/customization.adoc [hunk 12]
```diff
@@ -32,3 +32,133 @@ spec:
...
+[source,yaml]
+----
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: keycloak-app2-prod-config
+  namespace: bigli-test
```
Comment (tobru) [inline docs/modules/ROOT/pages/vshn-managed/keycloak/customization.adoc:140]:
```suggestion
  namespace: prod-app
```

File: docs/modules/ROOT/pages/vshn-managed/keycloak/customization.adoc [hunk 13]
```diff
@@ -32,3 +32,133 @@ spec:
...
+apiVersion: vshn.appcat.vshn.io/v1
+kind: VSHNKeycloak
+metadata:
+  name: keycloak-app2-prod
+  namespace: prod-app
+spec:
+  parameters:
+    service:
+      customConfigurationRef: keycloak-app2-prod-config # <1>
+      customEnvVariablesRef: keycloak-app2-prod-env # <2>
```
Comment (tobru) [inline docs/modules/ROOT/pages/vshn-managed/keycloak/customization.adoc:161]:
```suggestion
      customConfigurationRef: keycloak-app2-prod-config <1>
      customEnvVariablesRef: keycloak-app2-prod-env <2>
```

**Review by tobru — CHANGES_REQUESTED** [review]

File: docs/modules/ROOT/pages/vshn-managed/keycloak/customization.adoc (hunk 2 above)
Comment (TheBigLee) [reply docs/modules/ROOT/pages/vshn-managed/keycloak/customization.adoc]: Inventage will publish some docs on their end which documents also the default configuration they already apply and can be overwritten. This documentation will be amended once I've got their documentation.

File: docs/modules/ROOT/pages/vshn-managed/keycloak/customization.adoc (hunk 5 above)
Comment (TheBigLee) [reply docs/modules/ROOT/pages/vshn-managed/keycloak/customization.adoc:49]: This is a spring boot specific term: https://docs.spring.io/spring-boot/reference/features/external-config.html#features.external-config.files.configtree.

File: docs/modules/ROOT/pages/vshn-managed/keycloak/customization.adoc (hunk 2 above)
Comment (tobru) [reply docs/modules/ROOT/pages/vshn-managed/keycloak/customization.adoc]: Perfect, that makes sense. Thanks.

**Review by tobru — APPROVED** [review]
Please check the open suggestion, then I'm happy with the documentation. Thanks for the effort to incorporate my review.

**Pull request merged**